src/framework/pybind_json.hpp

```diff
@@ -271,4 +331,230 @@ void std::from_json(const json_t &js, py::object &o) {
 }
 
 //------------------------------------------------------------------------------
+
+//============================================================================
+// Pybind Conversion for Simulator types
+//============================================================================
+
+template<typename T> 
+py::object AERPy::from_matrix(const matrix<T> &mat) {
+  // THIS SHOULD RETURN A py::array_t but the author was la...
+  std::vector<std::vector<T> > tbr;
```

Comment:
As we know the final size of the 2d vector, we can use: `reserve()` to avoid allocations on every `push_back` , actually we can change `push_back` to `emplace_back` so we make sutere there's no temporary copies in the process.
```
tbr.reserve(rows);
```

Reply:
brilliant - changes made in ff4a6bc - even without a benchmark these changes are clearly beneficial

This will probably benefit the common cases more often but in a smaller way because these would only affect the statevector returned not the snapshots - so a constant factor benefit on a single 2^nqubits structure, rather than a constant factor on potentially many 2^nqubit structures

I'll keep any eye out for the opportunity to fix this in other places but currently I'm relying on pybind to do dynamic casting of vectors to lists (without copying them as I am essentially doing here)